mllib/src/test/scala/org/apache/spark/ml/regression/GeneralizedLinearRegressionSuite.scala

```diff
@@ -743,6 +743,84 @@ class GeneralizedLinearRegressionSuite
     }
   }
 
+  test("generalized linear regression with offset") {
+    /*
+      R code:
+      library(statmod)
+      df <- as.data.frame(matrix(c(
+        1.0, 1.0, 2.0, 0.0, 5.0,
+        2.0, 2.0, 0.5, 1.0, 2.0,
+        1.0, 3.0, 1.0, 2.0, 1.0,
+        2.0, 4.0, 0.0, 3.0, 3.0), 4, 5, byrow = TRUE))
+      families <- list(gaussian, poisson, Gamma, tweedie(1.5))
+      f1 <- V1 ~ -1 + V4 + V5
+      f2 <- V1 ~ V4 + V5
+      for (f in c(f1, f2)) {
+        for (fam in families) {
+          model <- glm(f, df, family = fam, weights = V2, offset = V3)
+          print(as.vector(coef(model)))
+        }
+      }
+
+      [1] 0.535040431 0.005390836
+      [1]  0.1968355 -0.2061711
+      [1]  0.307996 -0.153579
+      [1]  0.32166185 -0.09698986
+      [1] -0.8800000  0.7342857  0.1714286
+      [1] -1.9991044  0.7247511  0.1424392
+      [1] -0.27378146  0.31599396 -0.06204946
+      [1] -0.17118812  0.31200361 -0.02541656
+    */
+    val dataset = Seq(
+      OffsetInstance(1.0, 1.0, 2.0, Vectors.dense(0.0, 5.0)),
+      OffsetInstance(2.0, 2.0, 0.5, Vectors.dense(1.0, 2.0)),
+      OffsetInstance(1.0, 3.0, 1.0, Vectors.dense(2.0, 1.0)),
+      OffsetInstance(2.0, 4.0, 0.0, Vectors.dense(3.0, 3.0))
+    ).toDF()
+
+    val expected = Seq(
+      Vectors.dense(0.0, 0.535040431, 0.005390836),
+      Vectors.dense(0.0, 0.1968355, -0.2061711),
+      Vectors.dense(0.0, 0.307996, -0.153579),
+      Vectors.dense(0.0, 0.32166185, -0.09698986),
+      Vectors.dense(-0.88, 0.7342857, 0.1714286),
+      Vectors.dense(-1.9991044, 0.7247511, 0.1424392),
+      Vectors.dense(-0.27378146, 0.31599396, -0.06204946),
+      Vectors.dense(-0.17118812, 0.31200361, -0.02541656))
+
+    import GeneralizedLinearRegression._
+
+    var idx = 0
+    for (fitIntercept <- Seq(false, true)) {
+      for (family <- Seq("gaussian", "poisson", "gamma", "tweedie")) {
+        var trainer = new GeneralizedLinearRegression().setFamily(family)
```

Comment:
just a suggestion: maybe refactor the code below in a method and do fitIntercept.map(fi => family.map(fm => (fi, fm)).zip(expected).foreach(params => callMyMethod(params._1, params._2, params._3))
then you would get rid of the for loops and not have one long test case and you could remove the idx += 1 below